dm/syncer/syncer.go

```diff
@@ -1849,25 +1871,83 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
-		if shardingReSync == nil && len(shardingReSyncCh) > 0 {
-			// some sharding groups need to re-syncing
-			shardingReSync = <-shardingReSyncCh
-			savedGlobalLastLocation = lastLocation // save global last location
-			lastLocation = shardingReSync.currLocation
-
-			currentLocation = shardingReSync.currLocation
-			// if suffix>0, we are replacing error
-			s.isReplacingOrInjectingErr = currentLocation.Suffix != 0
-			s.locations.reset(shardingReSync.currLocation)
-			err = s.streamerController.ResetReplicationSyncer(s.runCtx, shardingReSync.currLocation)
-			if err != nil {
-				return err
-			}
+		if len(shardingReSyncCh) > 0 {
+			// for optimistic sharding mode, we must always accept new shardingResync and always use the newest sharding resync
+			// to handle the case that we finishes a shard lock before a shardingResync is finished, as the following case:
+			//                                                    newer
+			//   │                       ───────────────────────► time
+			//   │
+			//   │ tb1 conflict DDL1     │  ▲      │
+			//   │                       │  │      │
+			//   │       ...             │  │      │
+			//   │                       │  │      │
+			//   │ tb1 conflict DDL2     │  │      │  ▲     │
+			//   │                       │  │      │  │     │
+			//   │       ...             │  │      │  │     │
+			//   │                       │  │      │  │     │
+			//   │ tb2 conflict DDL1     ▼  │      │  │     │
+			//   │                                 │  │     │
+			//   │       ...           redirect    │  │     │
+			//   │                                 │  │     │
+			//   │ tb2 conflict DDL2               ▼  │     │
+			//   │                                          │
+			//   │       ...                     redirect   │
+			//   │                                          │
+			//   │  other dml events                        ▼
+			//   │
+			//   │                                       continue
+			//   ▼                                      replicating
+			//
+			// newer
+			// binlog
+			if (s.cfg.ShardMode == config.ShardPessimistic && shardingReSync == nil) || s.cfg.ShardMode == config.ShardOptimistic {
+				if s.cfg.ShardMode == config.ShardOptimistic && shardingReSync != nil {
+					unfinishedShardingResync := &ShardingReSync{
+						currLocation:   currentLocation,
+						latestLocation: shardingReSync.latestLocation.Clone(),
+						targetTable:    shardingReSync.targetTable.Clone(),
+						allResolved:    true,
+					}
+					s.tctx.L().Info("shardingResync is interrupted and pushed in shardingReSyncUnfinished",
+						zap.Stringer("interruptShardingResync", shardingReSync), zap.Stringer("pushedShardingResync", unfinishedShardingResync))
+					// save current location as the next time redirect position
+					shardingReSyncUnfinished = append(shardingReSyncUnfinished, unfinishedShardingResync)
+				}
+				// some sharding groups need to re-syncing
+				shardingReSync = <-shardingReSyncCh
```

Comment:
what if  `shardingReSync != nil`?
in old logic is `shardingReSync == nil && len(shardingReSyncCh) > 0 `

Reply:
There are two kind of situations:
Pessimistic: same with the old logic, only when `shardingReSync == nil && len(shardingReSyncCh) > 0` we can receive new `shardingReSync`.
Optimistic: we need to redirect again and immediately when we met new `shardingReSync`. This is to avoid the case that we finishes a shard lock before a shardingResync is finished, as written in the comment.

Reply:
sorry, this logic is so long that a whole screen is not displayed under 😂